konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/context/PipelineProfiler.java

```diff
@@ -0,0 +1,213 @@
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.FilenameUtils;
+import org.nd4j.common.io.StringUtils;
+import org.nd4j.common.primitives.AtomicBoolean;
+import org.nd4j.shade.jackson.databind.ObjectMapper;
+
+import java.io.*;
+import java.lang.management.ManagementFactory;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingDeque;
+
+@Slf4j
+public class PipelineProfiler implements Profiler {
+
+    private Writer writer;
+    private ObjectMapper json;
+    private final Thread fileWritingThread;
+    private final BlockingQueue<TraceEvent> writeQueue;
+    private final AtomicBoolean writing = new AtomicBoolean(false);
+
+    private long startTime;
+    private long endTime;
+    private final long pid;
+    private final long tid;
+
+    @Getter
+    private boolean logActive;
+    private ProfilerConfig profilerConfig;
+    private Path currentLog;
+
+    private long getProcessId() {
+        // Note: may fail in some JVM implementations
+        // therefore fallback has to be provided
+
+        // something like '<pid>@<hostname>', at least in SUN / Oracle JVMs
+        final String jvmName = ManagementFactory.getRuntimeMXBean().getName();
+        final int index = jvmName.indexOf('@');
+
+        if (index < 1) {
+            // part before '@' empty (index = 0) / '@' not found (index = -1)
+            return 0;
+        }
+
+        try {
+            return Long.parseLong(jvmName.substring(0, index));
+        } catch (NumberFormatException e) {
+            // ignore
+        }
+        return 0;
+    }
+
+    private void fileSizeGuard() throws IOException {
+        if ((profilerConfig.getSplitSize() > 0) &&
+            (Files.size(profilerConfig.getOutputFile()) > profilerConfig.getSplitSize())) {
+            String baseName = FilenameUtils.removeExtension(currentLog.getFileName().toString());
+            int num = 1;
+            String counted = org.apache.commons.lang3.StringUtils.EMPTY;
+            String[] parts = baseName.split("_");
+            if (parts.length > 1) {
+                counted = parts[0] + "_" + (Integer.parseInt(parts[1]) + 1);
+            }
+            else {
+                counted = baseName + "_" + num;
+            }
+            Path nextFile = Paths.get(currentLog.getParent() + FileSystems.getDefault().getSeparator() +
+                    counted + ".json");
+            Files.createFile(nextFile);
+            try {
+                this.writer = new BufferedWriter(new FileWriter(nextFile.toString(), true));
+                this.writer.write("[");     //JSON array open (array close is optional for Chrome profiler format)
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+            this.currentLog = nextFile;
```

Comment:
You should be flushing and closing the existing writer before creating the next one.